### ISSUE TITLE
Option to align the comment's opening and closing stars

### ISSUE THREAD

**Issue by mjambon** [OCamlPro/ocp-indent#312]
I didn't find an option to align the stars of closing comments. It's intended for comments that use a star at the beginning of each line, like this:
```ocaml
(* hello
 * world
 *)
```
The closing delimiter currently gets indented as follows by `ocp-indent`, which looks bad:
```ocaml
(* hello
 * world
*)
```
This happens regardless the `strict_comments` setting.

Option 1
--

I propose adding a setting for aligning the opening star with the closing star when the closing delimiter is on its own line. This would be done independently of what the comment contains. It would produce this:

```ocaml
(* hello
 * world
 *)

(*
 * hello
 * world
 *)

(* hello
   world
 *)

(*
   hello
   world
 *)

(* hello world *)

(* hello
   world *)
```

Option 2
--

Alternatively, if it's easy enough, we could only align the closing star if all the lines of the comment start with a star. This could be done by default without a special setting, and assume the old behavior is a bug. It would give this:
```
(*
   hello
*)

(*
 * hello
 *)

(*
 * hello
 * world
 *)

(*
   hello:
     * world
     * other
*)
```

Does it seem reasonable thing to add? Option 2 seems nicer for the user, being all automatic.

**Comment by NathanReb** [comment]
Sorry for the 6.5 years delay.

I started looking at this issue and I do believe option 2 is the best. I'm trying to understand all of the implications though, that means that the final `*)` would have its `*` aligned with the `(*` in the following cases:
1. strict_comment=false, all lines starts with a `*` already ALIGNED with the comment opening one, the comment close is on its own line
2. strict_comment=true, all lines starts with a `*`, the comment close is on its own line

I.e. the following:
```ocaml
(* hello
      * world
     *)
```
with strict_comments=false would be indented as follows:
```ocaml
(* hello
      * world
*)
```
and with strict_comments=true, as follows:
```ocaml
(* hello
 * world
 *)
```

**Comment by mjambon** [comment]
Thanks for looking into this! Your proposal looks right to me.

I was wondering about these comments:

(1) [prefer plain comments]
```ocaml
(* hello
*)
```
or

(2) [prefer starred comments]
```ocaml
(* hello
 *)
```
...because we don't know what the user wants by just looking at the comment. Depending on their preferred style, one example or the other might look broken to them.

I personally don't like starred comments because they require special editor support and in my opinion they don't improve readability. That's why I prefer (1) in a code base that's free of starred comments. However, we could also argue that both (1) and (2) are bad and that instead, the user should write one of these:

(3)
```ocaml
(* hello *)
```

(4)
```ocaml
(*
   hello
*)
```

(5)
```ocaml
(*
 * hello
 *)
```

Given that (3), (4), (5), are (universally?) better than (1) and (2) anyway, I think it's fine to just go with one or the other without a user setting and without any extra cleverness, when `strict_comments = true`.

**Comment by NathanReb** [comment]
Yes I think if it's ambiguous as in your example, I'll simply default back to the current indentation rather than to star aligned indentation.

Do you have any idea whether there should be an equivalent for doc comments?

**Comment by mjambon** [comment]
> Do you have any idea whether there should be an equivalent for doc comments?

I just ran a quick test on an existing project. The following code:
```ocaml
(** {1 Internal types}
 *
 *  These types are documented in the library's source code in [Types.ml]. They
 *  are subject to frequent and unannounced changes at the whim of the library's
 *  authors. A casual user should not need them.
 *)
```
is rendered by odoc as

<img width="718" height="144" alt="Image" src="https://github.com/user-attachments/assets/fcdf0f69-7383-45c4-bd73-a5fea4bd9a10" />

Based on this evidence, I would say that ocp-indent should treat stars within doc comments as ordinary text. Also, editors _should not_ auto-insert stars in doc comments but I didn't test this.

The comment above is currently (1.9.0, strict_comments = true) indented as:
```ocaml
(** {1 Internal types}
 *
 *  These types are documented in the library's source code in [Types.ml]. They
 *  are subject to frequent and unannounced changes at the whim of the library's
 *  authors. A casual user should not need them.
*)
```

I think it's fine but a bit misleading. I would consider not aligning the stars at all so that the misaligned `*)` doesn't look like a bug. I would consider this indentation (where a leading `*` is the same any text such as `x`):
```ocaml
(** {1 Internal types}
    *
    *  These types are documented in the library's source code in [Types.ml]. They
    *  are subject to frequent and unannounced changes at the whim of the library's
    *  authors. A casual user should not need them.
*)

```
This comment looks bad but it should be avoided by the user because the rendered documentation looks even worse. The idea is that the naive user will hopefully realize that something's wrong and that doc comments can't be starred comments.

Emitting a warning when finding such suspicious comments would be a good solution if ocp-indent had the ability of sending warnings to its caller. However, an IDE can do this easily without ocp-indent so it's not a big deal.

**Comment by NathanReb** [comment]
After looking at the implementation I realized ocp-indent doesn't have any special handling of doc comments and simply treats them as regular comments, hence the `strict_comments=true` behaviour you describe.

I left this untouched for now as it is somewhat unrelated to the support of star aligned comments but I agree having better support of doc comments would be for the best, starting with treating `*` in doc comments as just any regular character.
